src/main/kotlin/com/autonomousapps/tasks/ComputeAdviceTask.kt

```diff
@@ -254,10 +263,91 @@ internal class DependencyAdviceBuilder(
+     * them with added deps on their specific targets ("-jvm", "-android", etc). To solve this, we save off information
+     * about both in the below mappings and then cross-reference them in the second pass farther down.
+     *
+     * Finally, in the third pass, we do our usual processing of advice but with the cleaned KMP advice. It's
+     * important we clean up the KMP advice first so that the usual processing can continue to account for bundles
+     * in a simple way.
+     */
+
+    // Deferred KMP advice.
+    val deferredKmpAdvice = mutableListOf<Pair<Advice, Coordinates>>()
+    // Deferred non-KMP advice, we'll save these for the end.
+    val deferredNonKmpAdvice = mutableListOf<Pair<Advice, Coordinates>>()
+    // Track a mapping of configurations to kmp canonical deps that *have targets trying to add themselves*
+    // Note the values are just the canonical dep identifiers they correspond to, as the specific target isn't relevant
+    val addedKmpTargets = mutableMapOf<String, MutableSet<String>>()
+    // Track a mapping of removed KMP canonical deps. This is a mapping of configurations to the canonical dep identifier
+    val removedKmpCanonicalDeps = mutableMapOf<String, MutableSet<String>>()
+
+    dependencyUsages.asSequence()
       .flatMap { (coordinates, usages) ->
         StandardTransform(coordinates, declarations, supportedSourceSets, buildPath).reduce(usages).map { it to coordinates }
       }
+      .forEach { (advice, originalCoordinates) ->
+        when {
+          advice.isAdd() && originalCoordinates is ModuleCoordinates && originalCoordinates.isKmpTargetTarget -> {
+            deferredKmpAdvice += advice to originalCoordinates
+            addedKmpTargets.getOrPut(advice.toConfiguration!!, ::mutableSetOf).add(originalCoordinates.kmpCommonParentIdentifier())
+          }
+          advice.isRemove() && originalCoordinates is ModuleCoordinates && originalCoordinates.isKmpCanonicalDependency -> {
+            deferredKmpAdvice += advice to originalCoordinates
+            removedKmpCanonicalDeps.getOrPut(advice.fromConfiguration!!, ::mutableSetOf).add(originalCoordinates.identifier)
+          }
+          else -> {
+            deferredNonKmpAdvice += advice to originalCoordinates
+          }
+        }
+      }
+
+    val modifiedKmpAdvice = deferredKmpAdvice
+      // "null" removes the advice
+      .mapNotNull { (advice, originalCoordinates) ->
+        when {
+          // KMP target dependencies ("<dep>-android", "<dep>-jvm", etc) should defer to using their parent comment dep
+          // If the parent _isn't_ present in the original declarations though, then accept it and assume they're
+          // intentionally only depending on that specific target's artifact.
+          advice.isAdd() && originalCoordinates.isKmpTargetThatShouldDeferToParent(advice.toConfiguration, removedKmpCanonicalDeps) -> {
```

Comment:
You could use `advice.toConfiguration!!` since we know that's non-null here, and then simplify the method itself. If you're worried about it being _mis-used_, we could add an `error()` check. I'd prefer that for clarity vs the current null-check. As a possible alternative, you could pass in the advice itself, in which case the `!!` would be even more clear.

Reply:
[044613e](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/pull/919/commits/044613e44adcfd86a6f57abdb010bdbeaf596896)